homeassistant/util/template.py

```diff
@@ -0,0 +1,86 @@
+"""
+homeassistant.util.template
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Template utility methods for rendering strings with HA data.
+"""
+# pylint: disable=too-few-public-methods
+import json
+from jinja2.sandbox import ImmutableSandboxedEnvironment
+
+
+def render_with_possible_json_value(hass, template, value):
+    """ Renders template with value exposed.
+        If valid JSON will expose value_json too. """
+    variables = {
+        'value': value
+    }
+    try:
+        variables['value_json'] = json.loads(value)
+    except ValueError:
+        pass
+
+    return render(hass, template, variables)
+
+
+def render(hass, template, variables=None, **kwargs):
+    """ Render given template. """
+    if variables is not None:
+        kwargs.update(variables)
+
+    return ENV.from_string(template, {
```

Comment:
I thought about it adding it to render. I didn't do it because variables
that do not exist will be replaced with empty strings. I don't know what I
would return as a fallback for this method, the unprocessed template?

For the render_with_possible_json_value method I think logging the error
and returning the passed in value would be a good use case.

On Fri, Dec 11, 2015 at 10:20 AM, Philip Lundrigan <notifications@github.com
> wrote:

> In homeassistant/util/template.py
> <https://github.com/balloob/home-assistant/pull/720#discussion_r47386199>:
>
> > +        'value': value
> > +    }
> > +    try:
> > +        variables['value_json'] = json.loads(value)
> > +    except ValueError:
> > +        pass
> > +
> > +    return render(hass, template, variables)
> > +
> > +
> > +def render(hass, template, variables=None, **kwargs):
> > +    """ Render given template. """
> > +    if variables is not None:
> > +        kwargs.update(variables)
> > +
> > +    return ENV.from_string(template, {
>
> Should this be wrapped in some kind of exception handling? If the template
> string is incorrect, a jinja2.exceptions.UndefinedError is raised. This
> means anyone that uses the render or render_with_possible_json_value
> would have to import jinja2.exceptions if they want to handle the
> exception properly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/720/files#r47386199>.
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.